### PR TITLE
#721 refactor the crane infrastructure

### PR DESCRIPTION
**Pull request by saivirtue**
### Checklist
- [x] [Review Contribution Guidelines](https://github.com/oharastream/ohara/blob/master/CONTRIBUTING.md)
- [x] Add Reviewers (see [Authors](https://github.com/oharastream/ohara#authors))
- [x] Add Assignees (of course, you are the assignee by default!)
- [x] DON'T set a Milestone
- [x] DON'T add any labels (PR labels will be added automatically by Jenkins)
- [x] Link to an issue(You can do so by using the required by keyword: required by #721)
- [x] Add unit tests (or please explain why this PR is not worth having new tests)
- [ ] Pass QA (Sometimes you may encounter an existing flaky test which would obstruct you from getting QA passed. If the failed tests are unrelated, you can add a comment in the PR thread)

### CODE REVIEW

**Comment by saivirtue** [comment]
retry configurator

File: ohara-agent/src/main/scala/com/island/ohara/agent/docker/DockerCraneImpl.scala [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+      }
+  }
+
+  override protected def doClose(): Unit = {
+    Releasable.close(dockerCache)
+    Releasable.close(clusterCache)
+  }
+}
+
+object DockerCraneImpl {}
```
Comment (chia7712) [inline ohara-agent/src/main/scala/com/island/ohara/agent/docker/DockerCraneImpl.scala]: unused ?

File: ohara-agent/src/main/scala/com/island/ohara/agent/docker/DockerCraneImpl.scala [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+import com.island.ohara.client.configurator.v0.{ClusterInfo, StreamApi}
+import com.island.ohara.common.util.{Releasable, ReleaseOnce}
+import com.typesafe.scalalogging.Logger
+
+import scala.concurrent.{ExecutionContext, Future}
+
+private[agent] class DockerCraneImpl(nodeCollie: NodeCollie, dockerCache: DockerClientCache, executor: ExecutorService)
+    extends ReleaseOnce
+    with Crane {
+  private[this] val LOG = Logger("SshClusterCollie")
```
Comment (chia7712) [inline ohara-agent/src/main/scala/com/island/ohara/agent/docker/DockerCraneImpl.scala]: **LOG** is not a global member so lower case is more suitable.

File: ohara-agent/src/main/scala/com/island/ohara/agent/docker/ContainerCreator.scala [hunk 3]
```diff
@@ -142,7 +142,7 @@ trait ContainerCreator {
   /**
     * set docker container's network driver. implement by --network=$value
     *
-    * @param driver network driver
+    * @param networkDriver network driver
```
Comment (chia7712) [inline ohara-agent/src/main/scala/com/island/ohara/agent/docker/ContainerCreator.scala]: :)

File: ohara-agent/src/main/scala/com/island/ohara/agent/docker/StreamWarehouseImpl.scala [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+
+import scala.collection.JavaConverters._
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.Random
+
+private class StreamWarehouseImpl(nodeCollie: NodeCollie,
+                                  dockerCache: DockerClientCache,
+                                  clusterCache: Cache[Map[ClusterInfo, Seq[ContainerInfo]]])
+    extends StreamWarehouse {
+  private[this] val LOG = Logger(classOf[StreamWarehouse])
```
Comment (chia7712) [inline ohara-agent/src/main/scala/com/island/ohara/agent/docker/StreamWarehouseImpl.scala]: ditto

**Comment by saivirtue** [comment]
> OVERALL +1. How to integrate k8s impl in the future? (just curious)

I will try to create an entry in `Crane` for k8s. And for configurator-picture, `crane` will sit beside `clusterCollie`

**Pull request merged**